torchao/quantization/GPTQ.py

```diff
@@ -0,0 +1,579 @@
+        }
+
+        # trace model for one input
+        one_input = [multi.values[0] for multi in inputs]  # pyre-ignore[16]
+        exported_model = torch._dynamo.export(
+            model, aten_graph=True, pre_dispatch=True, tracing_mode="fake"
+        )(*one_input)
+        super().__init__(exported_model.graph_module)
+        self.new_state_dict = model.state_dict()
+        self.blocksize = blocksize
+        self.percdamp = percdamp
+        self.groupsize = groupsize
+        self.inputs = inputs
+        self.gptq_done = False
+        self.debug = False
+
+    def configure_quantization_mode(
+        self,
+        get_qparams_func,
+        quantize_func,
+        dequantize_func,
+        combine_qparams_list_func,
+        make_names_and_values_dict_func,
+        skip_layer_func,
+    ):
+        # these functions need to already be curried with all inputs other than weight, qparams
+        self.get_qparams_func = (
+            get_qparams_func  # accepts [2d weight tensor], outputs qparams.
+        )
+
+        self.quantize_func = quantize_func  # accepts [2d weight tensor], [qparams], outputs a 2d quantized tensor of desired dtype
+
+        self.dequantize_func = dequantize_func
+        # accepts [quantized] tensor and [qparams], outputs a 2d dequantized tensor of type float,
+        # assumes this output .to(w_orig_dtype) is ~eventual desired dequant behavior
+
+        self.combine_qparams_list_func = combine_qparams_list_func
+        # accepts [`list` of qparams] from quantizing one group at a time,
+        # outputs a qparams object that could be passed into quant/dequantize_func
+
+        self.skip_layer_func = skip_layer_func  # accepts [weight tensor], outputs a bool on whether or not to apply gptq to this layer
+
+        self.make_names_and_values_dict_func = make_names_and_values_dict_func  # accepts [2d quantized tensor], [qparams], returns a dict of names, values to put in state_dict
+        # note any final packing for storage should happen here
+        return self
+
+    def run(self):
+        assert (
+            self.get_qparams_func is not None
+        ), "need to configure quantization mode before running"
+        self.gptq_done = True
+        super().run(*self.inputs)
+
+    def get_quantized_state_dict(self):
+        assert (
+            self.gptq_done
+        ), "need to run GPTQRunner before you can get_quantized_state_dict"
+        quantized_state_dict = self.new_state_dict
+        # Don't want to store/load the kv_cache so remove it from the state_dict
+        del_list = []
+        for param_fqn in quantized_state_dict:
+            if "kv_cache" in param_fqn:
+                del_list.append(param_fqn)
+        for param_fqn in del_list:
+            quantized_state_dict.pop(param_fqn)
+        return quantized_state_dict
+
+    def call_function(self, target, args, kwargs, skip_quant=False):  # noqa: C901
```

Comment:
So is this also another way of replacing all linears with quantized linears?

Reply:
this is not replacement, this is part of an interpreter, only used to modify the runtime behavior of a captured graph.

but we could loop through the fx graph and replace ops: https://github.com/pytorch/examples/blob/main/fx/replace_op.py